source/extensions/filters/http/file_system_buffer/filter_config.cc

```diff
@@ -0,0 +1,202 @@
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "source/extensions/filters/http/file_system_buffer/filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace FileSystemBuffer {
+
+namespace {
+
+class BufferBehaviorStreamWhenPossible : public BufferBehavior {
+public:
+  static const BufferBehaviorStreamWhenPossible instance;
+};
+const BufferBehaviorStreamWhenPossible BufferBehaviorStreamWhenPossible::instance{};
+
+class BufferBehaviorBypass : public BufferBehavior {
+public:
+  static const BufferBehaviorBypass instance;
+  bool bypass() const override { return true; }
+};
+const BufferBehaviorBypass BufferBehaviorBypass::instance{};
+
+class BufferBehaviorInjectContentLengthIfNecessary : public BufferBehavior {
+public:
+  static const BufferBehaviorInjectContentLengthIfNecessary instance;
+  bool injectContentLength() const override { return true; }
+};
+const BufferBehaviorInjectContentLengthIfNecessary
+    BufferBehaviorInjectContentLengthIfNecessary::instance{};
+
+class BufferBehaviorFullyBufferAndAlwaysInjectContentLength : public BufferBehavior {
+public:
+  static const BufferBehaviorFullyBufferAndAlwaysInjectContentLength instance;
+  bool injectContentLength() const override { return true; }
+  bool replaceContentLength() const override { return true; }
+  bool alwaysFullyBuffer() const override { return true; }
+};
+const BufferBehaviorFullyBufferAndAlwaysInjectContentLength
+    BufferBehaviorFullyBufferAndAlwaysInjectContentLength::instance{};
+
+class BufferBehaviorFullyBuffer : public BufferBehavior {
+public:
+  static const BufferBehaviorFullyBuffer instance;
+  bool alwaysFullyBuffer() const override { return true; }
+};
+const BufferBehaviorFullyBuffer BufferBehaviorFullyBuffer::instance{};
+
+} // namespace
+
+// Not in the private namespace to facilitate testing of the NOT_SET path.
+const BufferBehavior& selectBufferBehavior(const ProtoBufferBehavior& behavior) {
+  switch (behavior.behavior_case()) {
+  case ProtoBufferBehavior::kStreamWhenPossible:
+    return BufferBehaviorStreamWhenPossible::instance;
+  case ProtoBufferBehavior::kBypass:
+    return BufferBehaviorBypass::instance;
+  case ProtoBufferBehavior::kInjectContentLengthIfNecessary:
+    return BufferBehaviorInjectContentLengthIfNecessary::instance;
+  case ProtoBufferBehavior::kFullyBufferAndAlwaysInjectContentLength:
+    return BufferBehaviorFullyBufferAndAlwaysInjectContentLength::instance;
+  case ProtoBufferBehavior::kFullyBuffer:
+    return BufferBehaviorFullyBuffer::instance;
+  case ProtoBufferBehavior::BEHAVIOR_NOT_SET:
+    // This should be impossible with the validate rule, but if it somehow happens we throw
+    // an exception.
+    break;
+  }
+  throw EnvoyException("invalid BufferBehavior in FileSystemBufferFilterConfig");
+}
+
+namespace {
+
+size_t getMemoryBufferBytesLimit(
+    const absl::InlinedVector<std::reference_wrapper<const ProtoStreamConfig>, 4>& configs) {
```

Comment:
How did you come up with 4 for all of these? I guess listener, virtual host, route or something? Either way can you maybe add some using aliases for all of these and add some comments? It would make it easier to read.

Reply:
Yeah, 4 was "listener, vhost, route, and one for futureproofing". Added alias and explanatory comment.